lightning/src/ln/peers/handshake/states.rs

```diff
@@ -1,30 +1,888 @@
+	}
+	pub(super) fn new_responder(responder_static_private_key: &SecretKey, responder_ephemeral_private_key: &SecretKey) -> Self {
+		HandshakeState::ResponderAwaitingActOne(ResponderAwaitingActOneState::new(responder_static_private_key.clone(), responder_ephemeral_private_key.clone()))
+	}
+}
+
+impl IHandshakeState for HandshakeState {
+	fn next(self, input: &[u8]) -> Result<(Option<Act>, HandshakeState), String> {
+		match self {
+			HandshakeState::InitiatorStarting(state) => { state.next(input) },
+			HandshakeState::ResponderAwaitingActOne(state) => { state.next(input) },
+			HandshakeState::InitiatorAwaitingActTwo(state) => { state.next(input) },
+			HandshakeState::ResponderAwaitingActThree(state) => { state.next(input) },
+			HandshakeState::Complete(_conduit) => { panic!("no acts left to process") }
+		}
+	}
+}
+
+// Handshake state of the Initiator prior to generating Act 1
+pub(super) struct InitiatorStartingState {
+	initiator_static_private_key: SecretKey,
+	initiator_static_public_key: PublicKey,
+	initiator_ephemeral_private_key: SecretKey,
+	initiator_ephemeral_public_key: PublicKey,
+	responder_static_public_key: PublicKey,
+	chaining_key: Sha256,
+	hash: Sha256
+}
+
+// Handshake state of the Responder prior to receiving Act 1
+pub(super) struct ResponderAwaitingActOneState {
+	responder_static_private_key: SecretKey,
+	responder_ephemeral_private_key: SecretKey,
+	responder_ephemeral_public_key: PublicKey,
+	chaining_key: Sha256,
+	hash: Sha256,
+	act_one_builder: ActBuilder
+}
+
+// Handshake state of the Initiator prior to receiving Act 2
+pub(super) struct InitiatorAwaitingActTwoState {
+	initiator_static_private_key: SecretKey,
+	initiator_static_public_key: PublicKey,
+	initiator_ephemeral_private_key: SecretKey,
+	responder_static_public_key: PublicKey,
+	chaining_key: ChainingKey,
+	hash: Sha256,
+	act_two_builder: ActBuilder
+}
+
+// Handshake state of the Responder prior to receiving Act 3
+pub(super) struct ResponderAwaitingActThreeState {
+	hash: Sha256,
+	responder_ephemeral_private_key: SecretKey,
+	chaining_key: ChainingKey,
+	temporary_key: [u8; 32],
+	act_three_builder: ActBuilder
+}
+
+impl InitiatorStartingState {
+	pub(crate) fn new(initiator_static_private_key: SecretKey, initiator_ephemeral_private_key: SecretKey, responder_static_public_key: PublicKey) -> Self {
+		let initiator_static_public_key = private_key_to_public_key(&initiator_static_private_key);
+		let (hash, chaining_key) = handshake_state_initialization(&responder_static_public_key);
+		let initiator_ephemeral_public_key = private_key_to_public_key(&initiator_ephemeral_private_key);
+		InitiatorStartingState {
+			initiator_static_private_key,
+			initiator_static_public_key,
+			initiator_ephemeral_private_key,
+			initiator_ephemeral_public_key,
+			responder_static_public_key,
+			chaining_key,
+			hash
+		}
+	}
+}
+
+impl IHandshakeState for InitiatorStartingState {
```

Comment:
I have to say, I'm a big fan of this concept. Moving the process_act_x() code to state advancement makes the flow much easier to follow and reason about.